### PR TITLE
Numerics fix from #40

### PR DESCRIPTION
**Pull request by ksivaman**
This pull request addresses the following 2 issues from PR #40 
- ~~Whether or not to compute dgrad is not determined solely by `input.requires_grad`. For a lot of the LLM cases this attribute is `False` but we still need to propagate the gradient.~~ The check for `inputmat.requires_grad` always returns `False` in the scenario of checking whether or not to compute dgrad because `inputmat` is the output of an operation and a new tensor of the same name. To do the check, we need to grab the `requires_grad` from `inp` instead. Since this needs to be done for all 3 APIs and is an orthogonal issue, it is out of the scope of this PR. 
- We update the forward scales/amaxes for the previous iteration in the beginning of the current iteration. Thus, `needs_stats` cannot be set only for training and calibration. Some key updates also happen on the first validation step.

Signed-off-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>

### CODE REVIEW

File: transformer_engine/pytorch/module.py [hunk 1]
```diff
@@ -417,23 +417,22 @@ def prepare_forward(
...
-                        )
-                        set_amax_buffer_key_deletion(self.fp8_meta, forward=True)
-                    else:
-                        amax_and_scale_update(
-                             self.fp8_meta, True, update_weight_scale_inv=update_weight_scale_inv
-                        )
+            update_weight_scale_inv = is_first_microbatch is None or is_first_microbatch
+
+            # Previous iteration was grad_enabled
+            if self.fp8_meta.get("update_amax_and_scale_fwd", False):
```
Comment (schetlur-nv) [inline transformer_engine/pytorch/module.py:423]: Is fp8_meta always guaranteed to exist here?

Comment (ksivaman) [reply transformer_engine/pytorch/module.py:423]: Yes, we initialize it in `TransformerEngineBaseModule`.

**Review by schetlur-nv — APPROVED** [review]

**Comment by ksivaman** [comment]
Internal ci passed
